### PR TITLE
fix: 1943 - FE - Investigate and fix random redirect from annotation page to /projects page

### PR DESCRIPTION
**Pull request by artikandri**
Issues fixed with this PR: 
Sometimes the page will redirect user randomly to the projects page. Our assumption for now is because the page is loaded before the apis because of poor internet connection.
https://redmine.e-science.pl/clarin/issues/1943 

Analysis: 
After doing some analysis, I decided that the problem indeed has something to do with the poor internet connection, but not with the `/index.vue` page, but rather the middleware. The `check-project` calls the `isStaff` value, but this value was being loaded after the `check-project` middleware loaded in first load, because the `check-project` middleware was loaded in the `nuxt.config.js` (prioritized) and the `check-auth` was loaded before loading each page (lower priority). Because of this, I moved the `check-auth` and `auth` to `config` and handled the code accordingly. 

Tested with throttled internet connection in Linux with this command: 
`sudo tc qdisc add dev lo root netem delay 100ms`
to remove: `sudo tc qdisc del dev lo root `

What's changed: 
- `frontend/layouts/*.vue`: removed `check-auth`, `auth`
- `frontend/middleware/auth.js`: added additional check to exlude outer pages from authorization 
- `frontend/middleware/check-auth.js`: added additional check to exlude outer pages from authorization 
- `frontend/nuxt.config.js`: added `check-auth`, `auth` middlewares
- `frontend/pages/projects/_id/affective-annotation/index.vue`: changed `isProjectAdmin` to `isStaff` in case the isProjectAdmin value wasnt properly loaded 
- `frontend/repositories/member/*.ts`: minor fix, handle empty/undefined id

### CODE REVIEW

**Review by teddy-f-47 — CHANGES_REQUESTED** [review]
Hello, thank you for the hard work! Unfortunately, when I test the code, now I encounter a different issue.

Based on my understanding, the original doccano implemented this "random redirect" in the first place as a fallback mechanism in case one of the APIs fails (e.g. due to slow internet connection like described in the PR desc). For example, if we open the annotation page and quickly navigate to somewhere else, the relation API may fail because at the moment when the request is made to the relation API the project id will be already undefined -- this results in 404 error. With the "random redirect", instead of showing the 404 error, it just redirects the user to the project list page. It appears "random" because the user doesn't know that one of the APIs failed.

Now that the redirect mechanism has been removed from the inner pages, if I try to navigate doccano very quickly, I often get 404 error, like in the screenshot below:
ssssss

To replicate the issue, try to run doccano in this branch, login, and just navigate as quickly as possible around project list page and annotation page. Because of this, while the proposed fix in this PR does eliminate the "random redirect" issue, I am not sure if it is an ideal solution, as now the user may get the 404 error instead of getting redirected.

**Comment by teddy-f-47** [comment]
I think we can/should keep your proposed fix. However, in addition, I suggest to add a loading animation that appears after opening one of the projects in the project list page (`pages\projects\index.vue`), and a loading animation that appears after opening the annotation page (`pages\projects\_id\affective-annotation\index.vue`). Such a loading animation that covers the whole screen can indirectly reduce the speed of the user in interacting/navigating doccano, and reduce the likelihood of encountering 404 error.

I know it does not really solve the root cause, but I think it is a simple solution that is pretty effective. 😅 Just to prevent the user from navigating way too quickly

edit: The loading screen does not need to wait for all APIs to finish (I think it can be a bit difficult to implement). I think it can appear for just like 3 seconds, for example.

**Comment by artikandri** [comment]
Hello, thank you for the feedback. Unfortunately, while of course adding the loading page shifts the user's attention, I can't agree with you because as you have stated it yourself, it doesnt solve the problem, and I will keep it as a last possible solution. I will think of how to handle the random redirect and go back here later, thank you.

**Comment by artikandri** [comment]
Hello @teddy-f-47 , I have updated the branch. However, I chose to handle the `undefined` ids instead of changing the default page, because default page is always supposed to be static, thats the selling point of `nuxt` instead of traditional rendering after all :) Can you check?

**Comment by teddy-f-47** [comment]
> Hello @teddy-f-47 , I have updated the branch. However, I chose to handle the `undefined` ids instead of changing the default page, because default page is always supposed to be static, thats the selling point of `nuxt` instead of traditional rendering after all :) Can you check?

Hello, thank you for the fix! I think the approach by handling each `undefined` id in each API request is great. Unfortunately, I still got the 404 error:
https://user-images.githubusercontent.com/64476430/220192302-94f1109b-0af0-47b7-99c0-f193d954f245.mp4
bug

 1. I have checked the problematic API requests, and at the very least, we need to add similar handling in `APIExampleRepository` (`frontend\repositories\example\apiDocumentRepository.ts`), inside two methods: the `async list` and the `async listStates`. Ideally we need to add similar handling in every API request that takes project id or example id as an input, but from my initial tests as a regular annotator navigating quickly inside doccano, adding the handling into these two methods seem to be enough to remove most if not all 404 error.
 2. A very minor thing, but I got a warning `Attribute "class" should go before "@click.once"` from `frontend\components\project\ProjectList.vue` line 36. 😅 

edit: Added gif version of the bug, because the mp4 video somehow didn't show up unless it is downloaded 😅

**Comment by artikandri** [comment]
Hello, thank you for your feedbacks. I have adjusted the last two apis to check the `projectId` and solved the `yarn errors`, can you check?

File: frontend/repositories/example/apiDocumentRepository.ts [hunk 1]
```diff
@@ -54,9 +58,13 @@ export class APIExampleRepository implements ExampleRepository {
...
   async listStates(projectId: string, exampleId: number): Promise<ExampleStateItemList> {
-    const url = `/projects/${projectId}/examples/${exampleId}/states`
-    const response = await this.request.get(url)
-    return plainToInstance(ExampleStateItemList, response.data)
+    let result = Promise.resolve({count: 0, next: null, prev: null, items: []})
+    if (String(projectId) !== "undefined" && String(exampleId) !== "undefined") {
+      const url = `/projects/${projectId}/examples/${exampleId}/states`
+      const response = await this.request.get(url)
+      result = response.data
+    }
```
Comment (teddy-f-47) [inline frontend/repositories/example/apiDocumentRepository.ts:66]: For some reason this line gives my an error on my machine, I have to change it to:
```
result = Promise.resolve({count: response.data.count, next: response.data.next, prev: response.data.previous, items: response.data.results})
```
I had to map those params explicitly in other methods as well, where there is `result = response.data`. But probably it is just a quirk on my windows machine. I sometimes also still got 404 errors from other endpoints: `http://localhost:3000/v1/projects/undefined/scale-types` and `http://localhost:3000/v1/projects/undefined/examples/601/spans` but they are quite rare.

**Review by teddy-f-47 — COMMENTED** [review]
Hello, thank you for the great work! In general it is ok now, there are still some quirks but maybe it is just my computer's issue.

**Review by teddy-f-47 — APPROVED** [review]

**Pull request closed without merging**